git: Use VolumeHost.GetExec() to execute stuff in volume plugins

jsafrane opened the pull request:
This prepares volume plugins to run things in containers instead of running
them on the host.

**Special notes for your reviewer**:

* instead of `cmd.SetDir(<dir>); cmd.Exec("git <command>")`, we do `cmd.Exec("git -C <dir> <command>")` - mounter.Exec does not have SetDir()
* there are lot of changes in the tests because a different exec interface is used.

@kubernetes/sig-storage-pr-reviews 

```release-note
gitRepo volumes in pods require git 1.8.5 or later
```

/assign @rootfs

jsafrane commented:
/sig storage

jsafrane commented:
/retest

rootfs commented on pkg/volume/git_repo/git_repo_test.go:
```diff
@@ -311,73 +278,36 @@ func doTestPlugin(scenario struct {
+	var commandLog []expectedCommand
+	execCallback := func(cmd string, args ...string) ([]byte, error) {
+		// command is 'git -C <dir> <command> <args>
+		gitCommand := args[2]
```
error out if len(args) is invalid

jsafrane replied:
good catch, fixed. Now I check that `arg[0] == "-C"`, just to be sure.

rootfs commented on pkg/volume/git_repo/git_repo_test.go:
```diff
@@ -311,73 +278,36 @@ func doTestPlugin(scenario struct {
+			}
 		}
+		// add the command to log with de-randomized gitDir
+		args[1] = strings.Replace(gitDir, mounter.GetPath(), "volume-dir", 1)
```
based on the context, it should just return `volume-dir` right?

jsafrane replied:
Yes, it *should* always return "volume-dir". However, if there is bug in the volume plugin it may be called with a different directory and the test should catch it.

rootfs replied:
why not just force it: `args[1] = "volume-dir"`, replace doesn't reset the entire string

jsafrane replied:
Because I want to check that correct `arg[1]` was used by the plugin. If the plugin uses `/var/lib/kubelet/pods/<uid>/kubernetes.io~git/foo/123` instead of correct `..kubernetes.io~git/foo/`, `Replace` will return `volume-plugin/123` and the test (correctly) fails.

jsafrane commented:
/retest

jsafrane commented:
/test all

jsafrane commented:
@rootfs not sure where we are with this, can it be merged?

rootfs commented:
/approve no-issue
/lgtm

rootfs commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane committed:
git: Use VolumeHost.GetExec() to execute stuff in volume plugins

This prepares volume plugins to run things in containers instead of running
them on the host.

jsafrane commented:
Fixed e2e tests (luckily we have them): I forgot to fill `exec` here: https://github.com/kubernetes/kubernetes/pull/51098/files#diff-45aabca84f5daab0bb5d8194a7282018R104

Which led to panic when `exec` was actually used here: https://github.com/kubernetes/kubernetes/pull/51098/files#diff-45aabca84f5daab0bb5d8194a7282018R250

@rootfs PTAL

rootfs commented:
/test pull-kubernetes-e2e-gce-etcd3

rootfs commented:
/test pull-kubernetes-e2e-gce-bazel

jsafrane commented:
```
UTC (durationBeforeRetry 500ms). Error: MountVolume.SetUp failed for volume "git-volume" (UniqueName: "kubernetes.io/git-repo/7b075d4e-8e4c-11e7-bd35-42010a800002-git-volume") pod "pod-secrets-7b0481e5-8e4c-11e7-a485-0a580a3c456b" (UID: "7b075d4e-8e4c-11e7-bd35-42010a800002") : failed to exec 'git clone http://10.0.32.88:2345 test':
Unknown option: -C
usage: git [--version] [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p|--paginate|--no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           [-c name=value] [--help]
           <command> [<args>]
: exit status 129
```

Damn

jsafrane commented:
The best way out of this is to add `exec.SetDir()`, just like `util/exec` does. But this cannot be easily implemented either in controller manager that will need to execute attach/detach utils via something like `kubectl exec` or in kubeler where CRI does not allow specification of working directory.

And I do not know about any smart way how to run `sh -c "cd <dir>; git <args>"` that would prevent shell from interpreting characters like  \`, $, ", ', ; and so on - part of args comes from user and must not be interpreted by shell and escaping these is dangerous.

@rootfs, any idea?

rootfs commented:
/keep-open

jsafrane commented:
/retest
(to see if the tests moved from Wheezy to GCI)

jsafrane commented:
/test e2e-gce

jsafrane commented:
@rootfs, it seems that the test env. was updated and we can merge this PR. I added a release note to use newer git.

rootfs commented:
/lgtm

jsafrane commented:
/test all

jsafrane commented:
/test kubernetes-cross

jsafrane commented:
/test pull-kubernetes-cross
